test/cpp/end2end/test_service_impl.cc

```diff
@@ -502,12 +502,20 @@ CallbackTestServiceImpl::ResponseStream(
       if (num_msgs_sent_ == server_responses_to_send_ - 1 &&
           server_coalescing_api_ != 0) {
         num_msgs_sent_++;
-        StartWriteLast(&response_, WriteOptions());
+        {
+          std::lock_guard<std::mutex> l(finish_mu_);
+          if (!finished_) {
+            StartWriteLast(&response_, WriteOptions());
+          }
+        }
         // If we use WriteLast, we shouldn't wait before attempting Finish
         FinishOnce(Status::OK);
       } else {
         num_msgs_sent_++;
```

Comment:
How is num_msgs_sent used? Do we want to increment it if we're not actually doing a write?

Reply:
At the current time we're only using it to drive NextWrite which won't get entered again, but I'll fix this for clarity.